keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -591,44 +592,45 @@ class SqlArtifactRepository(
       return false
     }
 
-    return selectPromotionReference(envUid, artUid, veto.version)
-      .fetchOne { (ref: String?) ->
-        ref?.let { reference ->
-          /**
-           * If there's a promotion reference, that means this artifact version was deployed as a result of
-           * another artifact version being vetoed. In that case, we don't veto unless [force] is enabled.
-           */
-          if (!force) {
-            log.warn(
-              "Not vetoing artifact version as it appears to have already been an automated rollback target: " +
-                "deliveryConfig=${deliveryConfig.name}, " +
-                "environment=${veto.targetEnvironment}, " +
-                "artifactVersion=${veto.version}, " +
-                "priorVersionReference=$reference")
-            return@fetchOne false
-          }
-        }
+    /**
+     * If there's a promotion reference, that means this artifact version was deployed as a result of
+     * another artifact version being vetoed. In that case, we don't veto unless [force] is enabled.
+     */
+    selectPromotionReference(envUid, artUid, veto.version)
+      .fetchOne() // null if [veto.version] never made it to the target environment
+      ?.value1() // null if the `promotion_reference` column is NULL
+      ?.let { reference ->
+      if (!force) {
```

Comment:
question: where are we sending this `force`? when can it be false?

Reply:
Looking through the code, I found two places:

* The [ApplicationController.veto] method, where we set `force=true`
* The [ResourceActuator.checkResource] method, where we don't pass a `force` argument at all, so it uses the default value in [ArtifactRepository.markAsVetoedIn], which is `false`.


[ApplicationController.veto]: https://github.com/spinnaker/keel/blob/master/keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/ApplicationController.kt#L201

[ResourceActuator.checkResource]:
https://github.com/spinnaker/keel/blob/master/keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/ApplicationController.kt#L201

[ArtifactRepository.markAsVetoedIn]:
https://github.com/spinnaker/keel/blob/master/keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepository.kt#L165

Reply:
Thanks! this is super helpful!